### PR TITLE
Add unique ID to Error class

### PR DESCRIPTION
**Pull request by a3d4**
This PR is supposed to finish assigning unique error IDs.

The next step is to add the IDs to the output.

### CODE REVIEW

File: liblangutil/Exceptions.h [hunk 1]
```diff
@@ -70,13 +81,15 @@ class Error: virtual public util::Exception
 	};
 
 	explicit Error(
+		ErrorId _errorId,
 		Type _type,
 		SourceLocation const& _location = SourceLocation(),
 		std::string const& _description = std::string()
 	);
```
Comment (a3d4) [inline liblangutil/Exceptions.h]: What is our policy regarding explicit constructors with two or more non-default arguments?

File: test/libsolidity/util/BytesUtils.cpp [hunk 2]
```diff
@@ -81,7 +81,7 @@ bytes BytesUtils::convertBoolean(string const& _literal)
 	else if (_literal == "false")
 		return bytes{false};
 	else
-		throw Error(Error::Type::ParserError, "Boolean literal invalid.");
+		throw Error(6991_error, Error::Type::ParserError, "Boolean literal invalid.");
```
Comment (a3d4) [inline test/libsolidity/util/BytesUtils.cpp]: Do we want unique IDs in tests?

Comment (chriseth) [reply test/libsolidity/util/BytesUtils.cpp]: I would say tests should not have IDs at all. Also, this is an exception and not an error that is added to the error list, isn't it?

**Comment by chriseth** [comment]
Needs rebase.

File: libsolidity/ast/Types.cpp [hunk 3]
```diff
@@ -173,7 +173,7 @@ void StorageOffsets::computeOffsets(TypePointers const& _types)
 			byteOffset = 0;
 		}
 		if (slotOffset >= bigint(1) << 256)
-			BOOST_THROW_EXCEPTION(Error(Error::Type::TypeError) << util::errinfo_comment("Object too large for storage."));
+			BOOST_THROW_EXCEPTION(Error(8708_error, Error::Type::TypeError) << util::errinfo_comment("Object too large for storage."));
```
Comment (a3d4) [inline libsolidity/ast/Types.cpp]: Perhaps we should report an error in a regular way. I plan to do it next. Then, hopefully, we'll have no more cases where `Error` is thrown.

File: test/libsolidity/util/BytesUtils.cpp (hunk 2 above)
Comment (a3d4) [reply test/libsolidity/util/BytesUtils.cpp]: Yes. I replaced it with another, test-specific exception (`TestParserError`, be83c54).

File: libsolidity/ast/Types.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/ast/Types.cpp]: I don't think it is possible here. I think we should use `solAssert` here.

File: liblangutil/Exceptions.cpp [hunk 4]
```diff
@@ -26,7 +26,10 @@ using namespace std;
 using namespace solidity;
 using namespace solidity::langutil;
 
-Error::Error(Type _type, SourceLocation const& _location, string const& _description):
+ErrorId solidity::langutil::operator"" _error(unsigned long long _error) { return ErrorId{ _error }; }
```
Comment (chriseth) [inline liblangutil/Exceptions.cpp]: Isn't this defined in the header already?

**Comment by chriseth** [comment]
Build error.

**Review by chriseth — APPROVED** [review]

File: libsolidity/ast/Types.cpp (hunk 3 above)
Comment (a3d4) [reply libsolidity/ast/Types.cpp]: OK, changed. But I would like to have a closer look later.

File: liblangutil/Exceptions.cpp (hunk 4 above)
Comment (a3d4) [reply liblangutil/Exceptions.cpp]: Now it is, c959341.

**Pull request merged**